File: python/src/skills_utils/doctor.py
from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from rich.console import Console

from skills_utils.agents import list_agents, resolve_skills_dir
from skills_utils.config_loader import home_dir, load_config, load_manifest, resolve_config_path
from skills_utils.github import GitHubClient
from skills_utils.repos import repo_from_config
from skills_utils.scanner import flat_symlink_path, nested_install_path
from skills_utils.sync import _ensure_flat_symlink

console = Console()


@dataclass
class CheckResult:
    level: str  # ok | warn | error
    message: str


def run_doctor(args: Any) -> int:
    config = load_config(args.config)
    manifest = load_manifest()
    results: list[CheckResult] = []

    config_path = resolve_config_path(args.config)
    if config_path.exists():
        results.append(CheckResult("ok", f"config.yaml found ({config_path})"))
    else:
        results.append(CheckResult("error", "config.yaml missing"))

    try:
        skill_count = sum(
            len(entries) for entries in manifest.get("agents", {}).values()
        )
        results.append(CheckResult("ok", f"manifest.json valid ({skill_count} skills tracked)"))
    except Exception as exc:
        results.append(CheckResult("error", f"manifest.json corrupt: {exc}"))

    token = args.token or GitHubClient._detect_token()
    if token:
        results.append(CheckResult("ok", "GitHub token available"))
    else:
        results.append(
            CheckResult(
                "warn",
                "No GitHub auth detected - set GITHUB_TOKEN/GH_TOKEN, pass --token, or run gh auth login (public repos OK)",
            )
        )

    with GitHubClient(token=token) as client:
        if client.ping():
            results.append(CheckResult("ok", "GitHub API reachable"))
        else:
            results.append(CheckResult("error", "GitHub API unreachable"))

        for slug, data in config.get("repos", {}).items():
            repo = repo_from_config(slug, data)
            try:
                dirs = client.list_skill_dirs(repo.owner, repo.repo, repo.skills_path)
                results.append(
                    CheckResult("ok", f"{repo.full_name} reachable ({len(dirs)} skill folders)")
                )
            except Exception as exc:
                results.append(CheckResult("error", f"{repo.full_name} unreachable: {exc}"))

    agents = list_agents(config)
    if args.agent:
        agents = [(aid, data) for aid, data in agents if aid == args.agent]

    for agent_id, agent in agents:
        skills_dir = resolve_skills_dir(agent)
        if skills_dir.exists() and os.access(skills_dir, os.W_OK):
            results.append(CheckResult("ok", f"{skills_dir} writable"))
        elif not skills_dir.exists():
            results.append(CheckResult("warn", f"{skills_dir} missing (creatable)"))
            if args.fix:
                skills_dir.mkdir(parents=True, exist_ok=True)
                results.append(CheckResult("ok", f"Created {skills_dir}"))
        else:
            results.append(CheckResult("error", f"{skills_dir} not writable"))

        agent_manifest = manifest.get("agents", {}).get(agent_id, {})
        for key, entry in agent_manifest.items():
            repo_slug = entry.get("repo_slug", key.split("/")[0])
            skill_name = key.split("/", 1)[-1]
            path = nested_install_path(skills_dir, repo_slug, skill_name)
            if not path.exists():
                results.append(CheckResult("warn", f"Orphan manifest entry: {agent_id}/{key}"))
                continue
            if not (path / "SKILL.md").exists():
                results.append(CheckResult("error", f"Missing SKILL.md: {path}"))

            link = flat_symlink_path(skills_dir, skill_name)
            if link.is_symlink():
                if link.exists():
                    results.append(CheckResult("ok", f"Symlink valid: {link.name}"))
                else:
                    results.append(CheckResult("error", f"Broken symlink: {link}"))
                    if args.fix:
                        link.unlink(missing_ok=True)
                        _ensure_flat_symlink(skills_dir, repo_slug, skill_name)
                        results.append(CheckResult("ok", f"Repaired symlink: {link.name}"))
            elif config.get("defaults", {}).get("flat_symlinks", True) and path.exists():
                if args.fix:
                    _ensure_flat_symlink(skills_dir, repo_slug, skill_name)
                    results.append(CheckResult("ok", f"Created symlink: {link.name}"))

    console.print("\n[bold]skills-sync doctor[/bold]\n")
    errors = warnings = 0
    for item in results:
        if item.level == "ok":
            console.print(f"  [green]✓[/green]  {item.message}")
        elif item.level == "warn":
            warnings += 1
            console.print(f"  [yellow]⚠[/yellow]  {item.message}")
        else:
            errors += 1
            console.print(f"  [red]✗[/red]  {item.message}")

    console.print()
    if errors:
        console.print(f"[red]{warnings} warnings, {errors} errors[/red]")
        return 2
    if warnings:
        console.print(f"[yellow]{warnings} warnings[/yellow]")
        return 1
    console.print("[green]All checks passed[/green]")
    return 0

File: node/src/doctor.ts
import { access, constants, lstat, mkdir } from "node:fs/promises";
import path from "node:path";
import { loadConfig, loadManifest, resolveConfigPath } from "./config.js";
import { GitHubClient } from "./github.js";
import { listAgents, resolveSkillsDir } from "./agents.js";
import { repoFromConfig } from "./repos.js";
import { nestedInstallPath } from "./scanner.js";
import { ensureFlatSymlink } from "./doctor-fix.js";

interface CheckResult {
  level: "ok" | "warn" | "error";
  message: string;
}

export async function runDoctor(args: any): Promise<number> {
  const config = await loadConfig(args.config);
  const manifest = await loadManifest();
  const results: CheckResult[] = [];
  const configPath = await resolveConfigPath(args.config);
  results.push({ level: "ok", message: `config.yaml found (${configPath})` });

  const skillCount = Object.values(manifest.agents ?? {}).reduce(
    (sum: number, entries: any) => sum + Object.keys(entries).length,
    0,
  );
  results.push({ level: "ok", message: `manifest.json valid (${skillCount} skills tracked)` });

  const token = args.token ?? (await GitHubClient.detectToken());
  if (token) results.push({ level: "ok", message: "GitHub token available" });
  else
    results.push({
      level: "warn",
      message:
        "No GitHub auth detected - set GITHUB_TOKEN/GH_TOKEN, pass --token, or run gh auth login (public repos OK)",
    });

  const client = new GitHubClient(token);
  if (await client.ping()) results.push({ level: "ok", message: "GitHub API reachable" });
  else results.push({ level: "error", message: "GitHub API unreachable" });

  for (const [slug, data] of Object.entries(config.repos ?? {})) {
    const repo = repoFromConfig(slug, data);
    try {
      const dirs = await client.listSkillDirs(repo.owner, repo.repo, repo.skillsPath);
      results.push({ level: "ok", message: `${repo.owner}/${repo.repo} reachable (${dirs.length} skill folders)` });
    } catch (error: any) {
      results.push({ level: "error", message: `${repo.owner}/${repo.repo} unreachable: ${error.message}` });
    }
  }

  let agents = listAgents(config);
  if (args.agent) agents = agents.filter(([id]) => id === args.agent);

  for (const [agentId, agent] of agents) {
    const skillsDir = await resolveSkillsDir(agent);
    try {
      await access(skillsDir, constants.W_OK);
      results.push({ level: "ok", message: `${skillsDir} writable` });
    } catch {
      try {
        await access(skillsDir);
        results.push({ level: "error", message: `${skillsDir} not writable` });
      } catch {
        results.push({ level: "warn", message: `${skillsDir} missing (creatable)` });
        if (args.fix) {
          await mkdir(skillsDir, { recursive: true });
          results.push({ level: "ok", message: `Created ${skillsDir}` });
        }
      }
    }

    for (const [key, entry] of Object.entries<any>(manifest.agents?.[agentId] ?? {})) {
      const repoSlug = entry.repo_slug ?? key.split("/")[0];
      const skillName = key.split("/").slice(1).join("/");
      const installPath = nestedInstallPath(skillsDir, repoSlug, skillName);
      try {
        await access(path.join(installPath, "SKILL.md"));
      } catch {
        results.push({ level: "warn", message: `Orphan manifest entry: ${agentId}/${key}` });
        continue;
      }
      const link = path.join(skillsDir, skillName);
      try {
        const info = await lstat(link);
        if (info.isSymbolicLink()) {
          try {
            await access(link);
            results.push({ level: "ok", message: `Symlink valid: ${skillName}` });
          } catch {
            results.push({ level: "error", message: `Broken symlink: ${link}` });
            if (args.fix) {
              const { rm, symlink } = await import("node:fs/promises");
              await rm(link, { force: true });
              await symlink(path.join(repoSlug, skillName), link);
              results.push({ level: "ok", message: `Repaired symlink: ${skillName}` });
            }
          }
        }
      } catch {
        if (args.fix && config.defaults?.flat_symlinks !== false) {
          await ensureFlatSymlink(skillsDir, repoSlug, skillName);
          results.push({ level: "ok", message: `Created symlink: ${skillName}` });
        }
      }
    }
  }

  console.log("\nskills-sync doctor\n");
  let errors = 0;
  let warnings = 0;
  for (const item of results) {
    const icon = item.level === "ok" ? "✓" : item.level === "warn" ? "⚠" : "✗";
    console.log(`  ${icon}  ${item.message}`);
    if (item.level === "warn") warnings += 1;
    if (item.level === "error") errors += 1;
  }
  console.log("");
  if (errors) {
    console.log(`${warnings} warnings, ${errors} errors`);
    return 2;
  }
  if (warnings) {
    console.log(`${warnings} warnings`);
    return 1;
  }
  console.log("All checks passed");
  return 0;
}
